infrastructure/src/main/java/org/corfudb/infrastructure/configuration/ServerConfigurator.java

```diff
@@ -0,0 +1,268 @@
+        managementServerRuntime = CorfuRuntime.fromParameters(corfuParams);
+        return managementServerRuntime;
+    }
+
+    private CorfuRuntime connect(CorfuRuntime runtime) {
+        Layout managementLayout = corfuConfig.getContext().copyManagementLayout();
+        if (managementLayout != null) {
+            managementLayout.getLayoutServers().forEach(runtime::addLayoutServer);
+        }
+        runtime.connect();
+
+        Runnable runtimeSystemDownHandler = () -> {
+            log.warn("ManagementServer: Runtime stalled. Invoking systemDownHandler after {} "
+                    + "unsuccessful tries.", runtime.getParameters().getSystemDownHandlerTriggerLimit());
+            throw new UnreachableClusterException("Runtime stalled. Invoking systemDownHandler after "
+                    + runtime.getParameters().getSystemDownHandlerTriggerLimit() + " unsuccessful tries.");
+        };
+
+        log.info("Corfu Runtime connected successfully");
+        runtime.getParameters().setSystemDownHandler(runtimeSystemDownHandler);
+        return runtime;
+    }
+
+    private SingletonResource<CorfuRuntime> getSingletonResourceRuntime() {
+        if (singletonResourceRuntime != null) {
+            return singletonResourceRuntime;
+        }
+
+        return SingletonResource.withInitial(() -> connect(getManagementServerRuntime()));
+    }
+
+    private HeartbeatCounter getHeartbeatCounter() {
+        if (heartbeatCounter != null) {
+            return heartbeatCounter;
+        }
+        heartbeatCounter = new HeartbeatCounter();
+        return heartbeatCounter;
+    }
+
+    private ClusterStateContext getClusterStateContext() {
+        if (clusterStateContext != null) {
+            return clusterStateContext;
+        }
+
+        ClusterState defaultView = ClusterState.builder()
+                .localEndpoint(corfuConfig.getContext().getLocalEndpoint())
+                .nodes(ImmutableMap.of())
+                .unresponsiveNodes(ImmutableList.of())
+                .build();
+        clusterStateContext = ClusterStateContext.builder()
+                .counter(getHeartbeatCounter())
+                .clusterView(new AtomicReference<>(defaultView))
+                .build();
+        return clusterStateContext;
+    }
+
+    private ManagementAgent getManagementAgent() {
+        if (managementAgent != null) {
+            return managementAgent;
+        }
+
+        FailureDetector failureDetector = new FailureDetector(getHeartbeatCounter(),
+                corfuConfig.getContext().getLocalEndpoint());
+
+        managementAgent = new ManagementAgent(
+                getSingletonResourceRuntime(),
+                corfuConfig.getContext(),
+                getClusterStateContext(),
+                failureDetector,
+                corfuConfig.getContext().copyManagementLayout());
+        return managementAgent;
+    }
+
+    private Orchestrator getOrchestrator() {
+        if (orchestrator != null) {
```

Comment:
`orchestrator` is never assigned, it's always null. Are you trying to make it singleton? I think it not necessary, since this will only be called once when bootstrapping server.

Reply:
It's the practice used in the DI frameworks like guice or spring when it comes to dependency injection. The instances are always cached since you never know when they will be reused. I can technically remove this statement.